[red-knot] Add initial support for `*` imports

AlexWaygood opened the pull request:
## Summary

This PR adds initial support for `*` imports to red-knot. The approach is to implement a standalone query, called from semantic indexing, that visits the module referenced by the `*` import and collects all global-scope public names that will be imported by the `*` import. The `SemanticIndexBuilder` then adds separate definitions for each of these names, all keyed to the same `ast::Alias` node that represents the `*` import.

There are many pieces of `*`-import semantics that are still yet to be done, even with this PR:
- This PR does not attempt to implement any of the semantics to do with `__all__`. (If a module defines `__all__`, then only the symbols included in `__all__` are imported, _not_ all public global-scope symbols.
- With the logic implemented in this PR as it currently stands, we sometimes incorrectly consider a symbol bound even though it is defined in a branch that is statically known to be dead code, e.g. (assuming the target Python version is set to 3.11):

  ```py
  # a.py

  import sys

  if sys.version_info < (3, 10):
      class Foo: ...

  ```

  ```py
  # b.py

  from a import *

  print(Foo)  # this is unbound at runtime on 3.11,
              # but we currently consider it bound with the logic in this PR
  ```

Implementing these features is important, but is for now deferred to followup PRs.

## Test Plan

Assertions in existing mdtests are adjusted, and several new ones are added.

TODO: the `collections.abc` integration test at the bottom of `star.md` is still failing. Not sure why...

AlexWaygood commented:
Still not sure why the `collections.abc` integration test is still failing, but other than that I'm pretty happy with this now!

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/definition.rs:
```diff
@@ -589,7 +657,27 @@ impl<'db> From<Option<Unpack<'db>>> for TargetKind<'db> {
+pub struct StarImportDefinitionKind {
+    node: AstNodeRef<ast::StmtImportFrom>,
+    symbol_id: ScopedSymbolId,
+    alias_index: usize,
```
It's sort-of unfortunate that we need to keep track of what index the alias is in the names list for `*` imports, since it's invalid syntax to do something like `from foo import *, X` -- if `*` is present in the names list of a `StmtImportFrom` node, it must be the _only_ name in the list. However, if we _don't_ keep track of the index the `*` is in the names list, we end up panicking on various kinds of invalid syntax.

MichaReiser replied:
Could we just search for the first alias with a * name? Or we could not  create a star import definition if the syntax is invalid

AlexWaygood replied:
Ohh of course, thanks! I'll make that change

AlexWaygood replied:
https://github.com/astral-sh/ruff/pull/16923/commits/a6f41525f2f0b9a175b6250a20c9319c72a40242

AlexWaygood commented:
> Still not sure why the `collections.abc` integration test is still failing, but other than that I'm pretty happy with this now!

I figured out the bug here and fixed it. Unfortunately, fixing the bug completely broke all our special casing for the builtin `len()` function. This is because `len()` has a parameter annotated with the `collections.abc.Sized` protocol in typeshed's `builtins.pyi` stub, and with my PR we now half-understand the collections.abc.Sized type (enough to emit false-positive errors 🙃) instead of not understanding it at all.

There were a huge amount of test failures in `len.md` resulting from this, so for now I've added a `KnownClass::Sized` variant and some special-casing for `KnownClass::Sized` to `Type::is_assignable_to()`. This is very similar to the existing special-casing @sharkdp had to add for `typing.SupportsIndex` in an earlier PR.

carljm commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+        }
+    }
+
+    fn visit_with_item(&mut self, with_item: &'db ast::WithItem) {
```
How much of the code in this visitor could we remove, without changing behavior, if we looked for all `Name` nodes with `Store` context?

I know imports don't use `Name` nodes and would still need dedicated handling (not to mention that we need to consider re-export logic). And I know we'd still need special handling for any node which creates a sub-scope, to avoid visiting the sub-scope. But maybe we wouldn't have to enumerate every possible place a target can appear?

AlexWaygood replied:
Ooh, interesting idea — I hadn't thought of that! Will try it out

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -4161,8 +4174,9 @@ impl<'db> FunctionType<'db> {
         let function_stmt_node = scope.node(db).expect_function();
-        let definition = semantic_index(db, scope.file(db)).definition(function_stmt_node);
-        Signature::from_function(db, definition, function_stmt_node)
+        let definitions = semantic_index(db, scope.file(db)).definitions(function_stmt_node);
```
I think it would be worth a helper method on the semantic index for getting a single definition, when that's what we expect, rather than repeating the debug assertion and indexing at every call site.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -2983,7 +2990,16 @@ impl<'db> TypeInferenceBuilder<'db> {
-            self.infer_definition(alias);
+            let definitions = self.index.definitions(alias);
+            if definitions.is_empty() {
+                // Ensure `from foo import *` is flagged as invalid if `foo` can't be resolved,
+                // *even if* `foo` is an empty module (and thus has no `Definition`s associated with it)
```
This comment doesn't quite make sense to me as written. If `foo` can't be resolved, then it can't be an "empty module" -- it's a nonexistent module. It seems like more what we want to say is, unresolved modules won't have any definitions associated with them, but we still need to emit a diagnostic for their unresolvability here.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3133,35 +3097,46 @@ impl<'db> TypeInferenceBuilder<'db> {
-            name,
-            asname: _,
-        } = alias;
+        let star_import_info =
```
Not sure why we need this `star_import_info` variable and two separate `if let`, when the variable is only used once? Why not just make this `let name = if let DefinitionKind::StarImport(star_import) = definition.kind(self.db()) {` and put all the logic in a single if-let?

AlexWaygood replied:
Borrow checker says no IIRC, but I'll double check!

AlexWaygood replied:
Yeah, so if I make this change:

```diff
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/src/types/infer.rs
index 4160a6aec..132001ec4 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -3161,21 +3161,16 @@ impl<'db> TypeInferenceBuilder<'db> {
             return;
         };
 
-        let star_import_info = definition
+        let name = definition
             .kind(self.db())
             .as_star_import()
             .map(|star_import| {
-                let symbol_table = self
-                    .index
-                    .symbol_table(self.scope().file_scope_id(self.db()));
-                (star_import, symbol_table)
-            });
-
-        let name = if let Some((star_import, symbol_table)) = star_import_info.as_ref() {
-            symbol_table.symbol(star_import.symbol_id()).name()
-        } else {
-            &alias.name.id
-        };
+                self.index
+                    .symbol_table(self.scope().file_scope_id(self.db()))
+                    .symbol(star_import.symbol_id())
+                    .name()
+            })
+            .unwrap_or(&alias.name.id);
```

then the borrow checker complains:

```
error[E0515]: cannot return value referencing temporary value
    --> crates/red_knot_python_semantic/src/types/infer.rs:3168:17
     |
3168 | //                 self.index
3169 | ||                     .symbol_table(self.scope().file_scope_id(self.db()))
     | ||________________________________________________________________________- temporary value created here
3170 | |                      .symbol(star_import.symbol_id())
3171 | |                      .name()
     | |____________________________^ returns a value referencing data owned by the current function
```

The borrow checker has to be able to prove that the value returned by `.symbol_table()` outlasts the `name` variable, since the `name` variable borrows from the value returned by `.symbol_table()`.

I'll add a comment saying why the indirection is necessary.

carljm reviewed (approved):
Nice work!

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+                self.possibly_add_export(&name.id);
+            }
+        } else {
+            self.possibly_add_export(&name.id);
```
argh, indeed! I actually added a test that was meant to check this... but there was a bug in the test 🙃

MichaReiser commented on crates/red_knot_python_semantic/src/module_name.rs:
```diff
@@ -234,3 +239,81 @@ impl std::fmt::Display for ModuleName {
+    Ok(module_name)
+}
+
+pub(crate) fn module_name_from_import_statement<'db>(
```
Nit: Should this be a method: `ModuleName::from_import_statement`

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+use crate::{module_name::module_name_from_import_statement, resolve_module, Db};
+
+#[salsa::tracked(return_ref)]
+pub(super) fn find_exports(db: &dyn Db, file: File) -> FxHashSet<Name> {
```
Nit: `find_exports` is a somewhat unusual name for a salsa query. We normally name these after what they return rather than using a verb. Maybe just `exports` or `exported_names`?

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+        if name.starts_with('_') {
+            return;
+        }
+        if !self.exports.contains(name) {
```
The `contains` + `insert` check has the downside that `name` gets hashed twice. I assume you do it this way to avoid cloning `name` when the name already exists (although that's probably unlikely?). https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.get_or_insert_with would be nice but it's nightly only.

I'm leaning towards always cloning. Or you can use a `FxHashMap<Name, ()>` with the entry API?

AlexWaygood replied:
Yeah, I thought you might have things to say about this ;)

I also wasn't sure if this was worth it. I switched it to just do the clone every time.

I tried the `FxHashMap` idea but as far as I can tell the `entry` API also requires the value passed in to be owned rather than be a reference? So I'm not sure that helps us here

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+impl<'db> Visitor<'db> for ExportFinder<'db> {
+    fn visit_alias(&mut self, alias: &'db ast::Alias) {
+        let ast::Alias { name, asname, .. } = alias;
+        if self.file.is_stub(self.db.upcast()) {
```
Nit: It may be worth to cache the `is_stub` lookup

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+            }) => {
+                self.visit_target(target);
+                self.visit_expr(iter);
+                for child in body {
+                    self.visit_stmt(child);
+                }
```
You can use `self.visit_body(body)` here

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -0,0 +1,299 @@
+    finder.exports
+}
+
+struct ExportFinder<'db> {
```
Not saying this is wrong, I'm just trying to understand the rational. 

What's the motivation for this being its own query vs. computing the exports as part of the semantic index builder?

AlexWaygood replied:
The trouble is that if `foo` has `from bar import *` in it, then we need to know the complete list of symbols exported from the global scope of `bar` in order to be able to complete semantic indexing of `foo`. I discussed this with @carljm and he was reluctant to have semantic indexing of one module be blocked on semantic indexing of another module, since semantic indexing can be expensive. (And the chain can go arbitrarily deep if `bar` also has `from baz import *` in it: now semantic indexing of `foo` is blocked on semantic indexing of `bar`, but semantic indexing of `bar` is blocked on semantic indexing of `baz`.)

Instead we decided to go down the route of a simpler, faster query called by semantic indexing that _only_ collects the names exported by the global scope. But the approach obviously has the disadvantage that we don't have all the tools available to us in this query that the `SemanticIndexBuilder` has.

MichaReiser replied:
Oh, I see. It recursively resolves star imports. That makes sense then

AlexWaygood replied:
It's my fault for not documenting this better -- I'll add a paragraph to the module doc-comment in `re_exports.rs` 👍

MichaReiser commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3133,35 +3097,46 @@ impl<'db> TypeInferenceBuilder<'db> {
-        } = alias;
+        let star_import_info =
+            if let DefinitionKind::StarImport(star_import) = definition.kind(self.db()) {
+                Some((star_import, symbol_table(self.db(), self.scope())))
```
```suggestion
                Some((star_import, self.index.symbol_table(self.scope())))
```

AlexWaygood replied:
hmm, that doesn't compile. This does compile:

```suggestion
                Some((
                    star_import,
                    self.index
                        .symbol_table(self.scope().file_scope_id(self.db())),
                ))
```

but it doesn't really seem simpler anymore ;-)

AlexWaygood replied:
Oh I see — I'll make the change, thanks

MichaReiser reviewed (commented):
This is great!

AlexWaygood commented:
(haven't finished addressing/responding to review comments yet, will continue tomorrow!)

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/import/star.md:
```diff
@@ -140,10 +130,163 @@ X = (Y := 3) + 4
+# error: [unresolved-reference]
+reveal_type(j)  # revealed: Unknown
+
+# TODO: these should all reveal `int`
+# (we don't generally model elsewhere in red-knot that bindings from walruses
+# "leak" from comprehension scopes into outer scopes, but we should)
+reveal_type(e)  # revealed: Unknown
+reveal_type(g)  # revealed: Unknown
+reveal_type(i)  # revealed: Unknown
+reveal_type(k)  # revealed: Unknown
```
mimimal repro that shows the bug here is pre-existing, rather than an issue with this PR: https://playknot.ruff.rs/6f9f0942-db9f-481f-b868-690be1d1176e

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -991,7 +1001,54 @@ where
+                            continue;
+                        };
+
+                        let Some(module) = resolve_module(self.db, &module_name) else {
+                            continue;
+                        };
```
I'm wondering if we should cache the resolved module in the `DefinitionKind` for `*` import definitions? It feels slightly silly to re-resolve the module name and module for these definitions in `infer.rs`, when we know that we wouldn't have created any definitions for the node during semantic indexing unless the module name and module were both resolvable.

AlexWaygood replied:
Hrm, this involves quite an annoying amount of refactoring unless we also cache the resolved `ModuleName` as well as the `File` on `StarImportDefinitionKind`. Leaning towards leaning this as-is for now, even though the duplication of logic between this file and `infer.rs` is somewhat annoying.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/import/star.md:
```diff
@@ -140,10 +130,168 @@ X = (Y := 3) + 4
+# error: [unresolved-reference]
+reveal_type(j)  # revealed: Unknown
+
+# TODO: these should all reveal `int`
+# (we don't generally model elsewhere in red-knot that bindings from walruses
+# "leak" from comprehension scopes into outer scopes, but we should)
+reveal_type(g)  # revealed: Unknown
+reveal_type(i)  # revealed: Unknown
+reveal_type(k)  # revealed: Unknown
+reveal_type(m)  # revealed: Unknown
```
mimimal repro that shows the bug here is pre-existing, rather than an issue with this PR: [playknot.ruff.rs/6f9f0942-db9f-481f-b868-690be1d1176e](https://playknot.ruff.rs/6f9f0942-db9f-481f-b868-690be1d1176e)

AlexWaygood replied:
Wroteup an issue for this: https://github.com/astral-sh/ruff/issues/16954

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -365,10 +372,8 @@ impl<'db> SemanticIndexBuilder<'db> {
-            .insert(definition_node.key(), definition);
-        debug_assert_eq!(existing_definition, None);
+        self.add_entry_for_definition_key(definition_node.key())
+            .push(definition);
```
I had a go at addressing this in https://github.com/astral-sh/ruff/pull/16923/commits/210b860387cb6333b3319f2ce814660eec70a0eb -- is that the kind of thing you were thinking of?

carljm replied:
Yep, that looks great!

carljm commented on crates/red_knot_python_semantic/resources/mdtest/import/star.md:
```diff
@@ -206,9 +206,71 @@ print((
+# error: [unresolved-reference]
+reveal_type(j)  # revealed: Unknown
+
+# TODO: these should all reveal `int`
```
probably `Unknown | int` since none of them are declared?

AlexWaygood replied:
Thanks, made this change in https://github.com/astral-sh/ruff/pull/16955

carljm commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1000,6 +1040,18 @@ where
+                        //
+                        // The reason why maintaining this invariant requires special handling here
+                        // is that some `Alias` nodes may be associated with 0 definitions:
+                        // - If the import statement has multiple `*` names in the `names` list
+                        //   (e.g. `from foo import *, bar, *`)
```
```suggestion
                        // - If the import statement has invalid syntax: multiple `*` names in the `names` list
                        //   (e.g. `from foo import *, bar, *`)
```

carljm commented on crates/red_knot_python_semantic/src/semantic_index/re_exports.rs:
```diff
@@ -3,6 +3,18 @@
+//!
+//! If we determined the set of exported names during semantic indexing rather than as a
+//! separate query, we would need to complete semantic indexing on `bar` in order to
+//! complete analysis of the global namespace of `foo`. Since semantic indexing is somewhat
```
Not sure if we have to say it here, but a main concern is also making semantic indexing potentially cyclic, meaning we'd have to fixpoint iterate it, which would be even more expensive.

AlexWaygood replied:
https://github.com/astral-sh/ruff/pull/16958

carljm reviewed (approved):
Looks good to me!

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/import/star.md:
````diff
@@ -40,8 +37,7 @@ reveal_type(X)  # revealed: Literal[42]
+reveal_type(X)  # revealed: bool
 ```
 
 ### Overridden by later definition
````
Thanks -- I'm tackling this in https://github.com/astral-sh/ruff/pull/16955. I'll rebase this PR on top of that branch
